src/main/java/com/example/shorturl/service/user/UserService.java

```diff
@@ -41,8 +49,49 @@ public UserEntity signUp(UserRequest userRequest) throws UserException {
         return userRepository.save(userEntity);
     }
 
+    public UserEntity adminCreateUser(AdminUserRequest userRequest) {
+
+        Set<RoleEntity> roles = userRequest.getRoles().size() == 0 ? ImmutableSet.of(roleRepository.findByName(ERole.ROLE_USER).get()) :
```

Comment:
I didn't get this code, could you please explain it?

Reply:
Yes.

It is possible that the admin in the request is missing some parameters.

If the "roles" are missing or it is empty the program adds the USER role by default;  otherwise, add those from the request. If the admin tries to add a role that doesn't exist, it throws an exception

Same case with the "password", if it is empy it creates a random password if not it uses the one from the request.

Then we build the userEntity. In this part the "id" is equal to null, if we save it in the database we will create a new row.

Then we check if there is an existing user with the same "userName" or "email". If there is one we update the "id" of the userEntity with the existing, if we save it in the database we will update a user because we are sending a specific "id"

If there is not an existing user with the same "userName" or "email" we just leave the "id" as null creating a new row in the database

Reply:
👍